CMake build broken with custom config using p256-m or everest

### Summary

When using CMake, the build fails when using a “3rdparty” feature (everest, p256-m) that is activated by any means other than editing `include/mbedtls/mbedtls_config.h`:

* because the build has a different `mbedtls/mbedtls_config.h` earlier in the include path;
* or the build defines `MBEDTLS_CONFIG_FILE`;
* or the build defines `MBEDTLS_USER_CONFIG_FILE`;
* or the options are enabled directly on the compiler command line.

### Details

Steps to reproduce:
```
mkdir build-p256m-cmake
cd build-p256m-cmake
CFLAGS='-Wall -Wextra -DMBEDTLS_PSA_CRYPTO_CONFIG -DMBEDTLS_P256M_EXAMPLE_DRIVER_ENABLED' cmake ..
```

Expectation: the build succeeds.

Actual result: a linking failure, because the files in `3rdparty/p256-m` are not built or linked.

```
… (no mention of 3rdparty) …
[ 20%] Building C object programs/aes/CMakeFiles/crypt_and_hash.dir/crypt_and_hash.c.o
[ 20%] Linking C executable crypt_and_hash
/usr/bin/ld: ../../library/libmbedcrypto.a(psa_crypto_driver_wrappers.c.o): in function `psa_driver_wrapper_sign_hash':
/home/gilpes01/z/dev/mbedtls/work/p256-m-zeroize/build-p256m-cmake/library/psa_crypto_driver_wrappers.c:319: undefined reference to `p256_transparent_sign_hash'
/usr/bin/ld: ../../library/libmbedcrypto.a(psa_crypto_driver_wrappers.c.o): in function `psa_driver_wrapper_verify_hash':
/home/gilpes01/z/dev/mbedtls/work/p256-m-zeroize/build-p256m-cmake/library/psa_crypto_driver_wrappers.c:423: undefined reference to `p256_transparent_verify_hash'
/usr/bin/ld: ../../library/libmbedcrypto.a(psa_crypto_driver_wrappers.c.o): in function `psa_driver_wrapper_generate_key':
/home/gilpes01/z/dev/mbedtls/work/p256-m-zeroize/build-p256m-cmake/library/psa_crypto_driver_wrappers.c:854: undefined reference to `p256_transparent_generate_key'
/usr/bin/ld: ../../library/libmbedcrypto.a(psa_crypto_driver_wrappers.c.o): in function `psa_driver_wrapper_key_agreement':
/home/gilpes01/z/dev/mbedtls/work/p256-m-zeroize/build-p256m-cmake/library/psa_crypto_driver_wrappers.c:2849: undefined reference to `p256_transparent_key_agreement'
collect2: error: ld returned 1 exit status
make[2]: *** [programs/aes/CMakeFiles/crypt_and_hash.dir/build.make:136: programs/aes/crypt_and_hash] Error 1
make[1]: *** [CMakeFiles/Makefile2:966: programs/aes/CMakeFiles/crypt_and_hash.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
```

The problem is visible in `3rdparty/CMakeLists.txt`: the subdirectories are only traversed if
```
execute_process(COMMAND ${MBEDTLS_PYTHON_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/../scripts/config.py -f ${CMAKE_CURRENT_SOURCE_DIR}/../include/mbedtls/mbedtls_config.h get MBEDTLS_ECDH_VARIANT_EVEREST_ENABLED RESULT_VARIABLE everest_result)
execute_process(COMMAND ${MBEDTLS_PYTHON_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/../scripts/config.py -f ${CMAKE_CURRENT_SOURCE_DIR}/../include/mbedtls/mbedtls_config.h get MBEDTLS_P256M_EXAMPLE_DRIVER_ENABLED RESULT_VARIABLE p256m_result)
```

This only works if the configuration is the one in `include/mbedtls/mbedtls_config.h`. It doesn't work when the respective options are enabled because the build has a different `mbedtls/mbedtls_config.h` earlier in the include path, or the build defines `MBEDTLS_CONFIG_FILE`,  or the build defines `MBEDTLS_USER_CONFIG_FILE`, or the options are enabled directly on the compiler command line.

A tempting fix is to traverse the `3rdparty` directories unconditionally, but I don't know if that causes some breakage (why didn't we do that in the first place?).

--- a/3rdparty/CMakeLists.txt
+++ b/3rdparty/CMakeLists.txt
@@ -1,10 +1,2 @@
-execute_process(COMMAND ${MBEDTLS_PYTHON_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/../scripts/config.py -f ${CMAKE_CURRENT_SOURCE_DIR}/../include/mbedtls/mbedtls_config.h get MBEDTLS_ECDH_VARIANT_EVEREST_ENABLED RESULT_VARIABLE everest_result)
-execute_process(COMMAND ${MBEDTLS_PYTHON_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/../scripts/config.py -f ${CMAKE_CURRENT_SOURCE_DIR}/../include/mbedtls/mbedtls_config.h get MBEDTLS_PSA_P256M_DRIVER_ENABLED RESULT_VARIABLE p256m_result)
-
-if(${everest_result} EQUAL 0)
-    add_subdirectory(everest)
-endif()
-
-if(${p256m_result} EQUAL 0)
-    add_subdirectory(p256-m)
-endif()
+add_subdirectory(everest)
+add_subdirectory(p256-m)

--- a/3rdparty/everest/CMakeLists.txt
+++ b/3rdparty/everest/CMakeLists.txt
@@ -11,6 +11,19 @@ target_include_directories(everest
           include/everest/kremlib
           ${MBEDTLS_DIR}/library/)
 
+# Pass-through MBEDTLS_CONFIG_FILE and MBEDTLS_USER_CONFIG_FILE
+# This must be duplicated from library/CMakeLists.txt because
+# everest is not directly linked against any mbedtls targets
+# so does not inherit the compile definitions.
+if(MBEDTLS_CONFIG_FILE)
+    target_compile_definitions(everest
+        PUBLIC MBEDTLS_CONFIG_FILE="${MBEDTLS_CONFIG_FILE}")
+endif()
+if(MBEDTLS_USER_CONFIG_FILE)
+    target_compile_definitions(everest
+        PUBLIC MBEDTLS_USER_CONFIG_FILE="${MBEDTLS_USER_CONFIG_FILE}")
+endif()
+
 if(INSTALL_MBEDTLS_HEADERS)
 
   install(DIRECTORY include/everest

--- a/3rdparty/p256-m/CMakeLists.txt
+++ b/3rdparty/p256-m/CMakeLists.txt
@@ -9,6 +9,19 @@ target_include_directories(p256m
          $<INSTALL_INTERFACE:include>
   PRIVATE ${MBEDTLS_DIR}/library/)
 
+# Pass-through MBEDTLS_CONFIG_FILE and MBEDTLS_USER_CONFIG_FILE
+# This must be duplicated from library/CMakeLists.txt because
+# p256m is not directly linked against any mbedtls targets
+# so does not inherit the compile definitions.
+if(MBEDTLS_CONFIG_FILE)
+    target_compile_definitions(p256m
+        PUBLIC MBEDTLS_CONFIG_FILE="${MBEDTLS_CONFIG_FILE}")
+endif()
+if(MBEDTLS_USER_CONFIG_FILE)
+    target_compile_definitions(p256m
+        PUBLIC MBEDTLS_USER_CONFIG_FILE="${MBEDTLS_USER_CONFIG_FILE}")
+endif()
+
 if(INSTALL_MBEDTLS_HEADERS)
 
   install(DIRECTORY :${CMAKE_CURRENT_SOURCE_DIR}

--- a/ChangeLog.d/fix-cmake-3rdparty-custom-config.txt
+++ b/ChangeLog.d/fix-cmake-3rdparty-custom-config.txt
@@ -1,0 +1,3 @@
+Bugfix
+   * Fix the build with CMake when Everest or P256-m is enabled through
+     a user configuration file or the compiler command line. Fixes #8165.

